File: 0946 validate-stack-sequences/solution.cpp
class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        stack<int> st;
        int top1 = 0;
        for(int i : pushed) {
            st.push(i);
            while (!st.empty() && top1 < popped.size() && popped[top1] == st.top()) {
                st.pop();
                top1++;
            }
        }
        return st.empty() && top1 == popped.size();
    }
};

File: 0946 validate-stack-sequences/solution.java
class Solution {
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        int[] st = new int[pushed.length];
        int top = -1, top1 = 0;
        for (int i : pushed) {
            st[++top] = i;
            while (top >=0 && top1 < popped.length && popped[top1] == st[top]) {
                top--;
                top1++;
            }
        }
        return top ==-1 && top1 == popped.length;
    }
}
